src/java.base/share/classes/jdk/internal/classfile/Opcode.java

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.classfile;
+
+import java.lang.constant.ConstantDesc;
+import java.lang.constant.ConstantDescs;
+
+/**
+ * Describes the opcodes of the JVM instruction set, as well as a number of
+ * pseudo-instructions that may be encountered when traversing the instructions
+ * of a method.
+ *
+ * @see Instruction
+ * @see PseudoInstruction
+ */
+public enum Opcode {
+    NOP(Classfile.NOP, 1, Kind.NOP),
```

Comment:
This also duplicates the constants in classfile...

Reply:
On the contrary, it has been deduplicated. Opcode is referencing numeric constants stored in Classfile.

Reply:
sure, but my question is - once you have a nice enum that is 1-1 with the opcodes - why would a client want to use the low-level opcode bytes? Shouldn't the API only deal with Opcodes? (and, in the rare occurrence where a client wants to really know the int value of an opcode, they can do e.g. `PUTSTATIC.bytecode()`)

Reply:
It is the question of co-location of all numeric constants in one place in Classfile (or not).
I would agree with your proposal if the architectural decision is to decentralise the numeric constants and reduce Classfile footprint.